src/Codec/LazyBSONArrayCodec.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+namespace MongoDB\Codec;
+
+use MongoDB\BSON\PackedArray;
+use MongoDB\Exception\UnexpectedValueException;
+use MongoDB\Model\LazyBSONArray;
+
+use function array_values;
+use function sprintf;
+
+/**
+ * Codec for lazy decoding of BSON PackedArray instances
+ *
+ * @template-implements Codec<PackedArray, LazyBSONArray>
+ */
+class LazyBSONArrayCodec implements Codec, KnowsCodecLibrary
+{
+    /** @var CodecLibrary|null */
+    private $library = null;
+
+    public function attachLibrary(CodecLibrary $library): void
+    {
+        $this->library = $library;
+    }
+
+    /** @inheritDoc */
+    public function canDecode($value): bool
+    {
+        return $value instanceof PackedArray;
+    }
+
+    /** @inheritDoc */
+    public function canEncode($value): bool
+    {
+        return $value instanceof LazyBSONArray;
+    }
+
+    /** @inheritDoc */
+    public function decode($value): LazyBSONArray
+    {
+        if (! $value instanceof PackedArray) {
+            throw new UnexpectedValueException(sprintf('"%s" can only decode from "%s" instances', self::class, PackedArray::class));
+        }
+
+        return new LazyBSONArray($value, $this->getLibrary());
+    }
+
+    /** @inheritDoc */
+    public function encode($value): PackedArray
+    {
+        if (! $value instanceof LazyBSONArray) {
+            throw new UnexpectedValueException(sprintf('"%s" can only encode "%s" instances', self::class, LazyBSONArray::class));
+        }
+
+        $return = [];
+        foreach ($value as $offset => $offsetValue) {
+            $return[$offset] = $this->getLibrary()->canEncode($offsetValue) ? $this->getLibrary()->encode($offsetValue) : $offsetValue;
+        }
+
+        return PackedArray::fromPHP(array_values($return));
+    }
+
+    private function getLibrary(): CodecLibrary
+    {
+        return $this->library ?? $this->library = new CodecLibrary(new LazyBSONDocumentCodec(), $this);
```

Comment:
Originally I wanted to use a null-coalescing assignment, but those aren't available in 7.2. Your assumption however is correct: in the absence of a codec library (which we assume will be attached to the codec earlier) we create our own codec library consisting of the two codecs necessary to handle the lazy objects. I can refactor this if you'd like.

Reply:
The assignment within a `return` statement is really the thing that rubbed me the wrong way. That feels a bit like doing assignments within conditionals, which we also try to avoid.

I wouldn't object to:

```
$this->library = $this->library ?? new CodecLibrary(new LazyBSONDocumentCodec(), $this);
```

Once we require PHP 7.4+, I expect the CS tools would suggest a refactoring to use `??=`.